docs/content/guides/dagster/pythonic-resources.mdx

```diff
@@ -0,0 +1,337 @@
+
+Resources can be configured using environment variables, which is useful for secrets or other environment-specific configuration.
+
+To use environment variables, pass an <PyObject object="EnvVar" /> when constructing your resource.
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resources_env_vars endbefore=end_new_resources_env_vars
+from dagster._config.field_utils import EnvVar
+from dagster import Definitions
+from dagster._config.structured_config import ConfigurableResource
+
+
+class CredentialsResource(ConfigurableResource):
+    username: str
+    password: str
+
+
+defs = Definitions(
+    assets=...,
+    resources={
+        "credentials": CredentialsResource(
+            username=EnvVar("MY_USERNAME"),
+            password=EnvVar("MY_PASSWORD"),
+        )
+    },
+)
+```
+
+---
+
+## Configuring resources at runtime
+
+In some cases, you may want to specify configuration for a resource at runtime, in the launchpad. For example, you may want to specify a different target table in a database resource for each run.
+
+To do this, you can use the `configure_at_launch()` method to defer the construction of a configurable resource until runtime.
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resource_runtime endbefore=end_new_resource_runtime
+class DatabaseResource(ConfigurableResource):
+    table: str
+
+    def read(self):
+        ...
+
+
+@asset
+def data_from_database(db_conn: DatabaseResource):
+    return db_conn.read()
+
+
+defs = Definitions(
+    assets=...,
+    resources={"db_conn": DatabaseResource.configure_at_launch()},
+)
+```
+
+---
+
+## Resources which depend on other resources
+
+In some situations, you may want to define a resource which depends on other resources. For example, a resource representing a connection to an external service may rely on a resource representing credentials for that service.
+
+In this case, you can list that nested resource as an attribute of your resource class.
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resources_nesting endbefore=end_new_resources_nesting
+from dagster import Definitions
+from dagster._config.structured_config import ConfigurableResource
+
+
+class CredentialsResource(ConfigurableResource):
+    username: str
+    password: str
+
+
+class FileStoreBucket(ConfigurableResource):
+    credentials: CredentialsResource
+    region: str
+
+    def write(self, data: str):
+        get_filestore_client(
+            username=self.credentials.username,
+            password=self.credentials.password,
+            region=self.region,
+        ).write(data)
+
+
+credentials = CredentialsResource(username="my_user", password="my_password")
+defs = Definitions(
+    assets=...,
+    resources={
+        "bucket": FileStoreBucket(
+            credentials=credentials,
+            region="us-east-1",
+        ),
+    },
```

Comment:
We should decide on best practice in terms of whether or not we encourage people to a set a top-level resource for something like credentials here, even if we do support it in this style.